accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+
+The first .NET runtime that included it was .NET Framework 2.0, and the implementation has not changed much since then. The mitigation in C/C++ compilers have seen improvements in the meantime (for example, [Enhanced GS](https://msrc-blog.microsoft.com/2009/03/20/enhanced-gs-in-visual-studio-2010/) or -[fstack-protector-strong](https://lwn.net/Articles/584225/)).
+
+We will review and update stack buffer overflow protection in .NET runtime to match state of the art.
+
+# Language, Libraries and Runtime Features
+
+Library and language design have always played a key role in mitigating security bugs by preventing or discouraging them from being introduced in the first place.
+
+## Unsafe API and C# language constructs
+
+Number of newer APIs such as `Memory<T>` and friends come with security risks that are not clearly communicated.
+
+We need to adopt a more holistic approach to unsafe features and APIs that we expose for public use. Developers need feedback in the IDE so that they understand that they are using constructs that are unsafe and are [informed of the specific risks that come with using those APIs](https://github.com/dotnet/runtime/issues/41418). Over time, we bias more to constructs and paradigms that are safe by default (possibly inspired by Rust).
+
+## Object and Array Pooling
+
+`ArrayPool<T>` and other similar custom object pools are equivalent of the classic C malloc. Manual lifetime management traditionally suffers from use-after-free and double-free security vulnerabilities. Many years were spent on implementing diagnostics and mitigations of these vulnerabilities for malloc.
+
+We need to implement similar diagnostic and mitigation capabilities for ArrayPool\&lt;T\&gt; and friends. It is [not unusual](https://github.com/dotnet/corefx/pull/37270) to find security bugs caused by incorrect `ArrayPool<T>` use even in dotnet/runtime libraries.
+
+We are also going to be continually pushed to expose APIs like those we use internally which make potential misuse even easier, e.g. `ValueStringBuilder`. Thus far we have simply said &quot;we're not comfortable doing that&quot;, but it is unlikely going to be a satisfactory answer long-term. It is something we need to keep thinking about.
+
+## Secrets in Presence of GC Heap Compaction
+
+GC heap compaction can create a second copy of secrets stored in managed memory. It makes it impossible for an application to reliably clear in-memory secrets stored in managed memory after they are no longer used.
+
+We will introduce GC feature that clears unused memory after GC heap compaction. Having the GC clear memory after compaction improves the security of applications for which heap dumps and in-memory probes are within their scope of threat. [dotnet/runtime#10480](https://github.com/dotnet/runtime/issues/10480) has more details.
```

Comment:
Nit: "We will introduce _a_ GC feature"

Reply:
Just curious, did we ever estimate what sort of perf hit this would imply? Would it need to be opt-in?

Reply:
I think I heard a 5% perf hit \[to GC\] being thrown around somewhere? Every discussion I've had on this feature assumes customers would opt-in to it.